argocd-repo-server crashed

**Describe the bug**

Repo server was continuously crashing with the error below:

```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x1f709c6]
goroutine 1252 [running]:
github.com/argoproj/argo-cd/v2/reposerver/apiclient.(*ManifestResponse).MarshalToSizedBuffer(0x0, 0x4a4f6e8, 0x0, 0x0, 0x40bb45, 0x29999e0, 0x2c52420)
	/go/src/github.com/argoproj/argo-cd/reposerver/apiclient/repository.pb.go:2043 +0x26
github.com/argoproj/argo-cd/v2/reposerver/apiclient.(*ManifestResponse).Marshal(0x0, 0x2c52420, 0x0, 0x7fd9fc3f6198, 0x0, 0x1)
	/go/src/github.com/argoproj/argo-cd/reposerver/apiclient/repository.pb.go:2026 +0x7a
google.golang.org/grpc/encoding/proto.codec.Marshal(0x2c52420, 0x0, 0x0, 0x0, 0x40a51f, 0xc00003a000, 0x29b3380)
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/encoding/proto/proto.go:70 +0x199
google.golang.org/grpc.encode(0x7fd9fc4b8450, 0x4a4f6e8, 0x2c52420, 0x0, 0x4a4f6e8, 0x321f6f8, 0xc001fd12c0, 0x2ca6c80, 0xc000e64100)
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/rpc_util.go:511 +0x52
google.golang.org/grpc.(*Server).sendResponse(0xc000402600, 0x3242c38, 0xc0014ad380, 0xc000e64000, 0x2c52420, 0x0, 0x0, 0x0, 0xc001c10f9a, 0x0, ...)
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:805 +0x91
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000402600, 0x3242c38, 0xc0014ad380, 0xc000e64000, 0xc00055e510, 0x48f47c0, 0x0, 0x0, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:1004 +0x5d1
google.golang.org/grpc.(*Server).handleStream(0xc000402600, 0x3242c38, 0xc0014ad380, 0xc000e64000, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:1208 +0x122c
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000544630, 0xc000402600, 0x3242c38, 0xc0014ad380, 0xc000e64000)
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:686 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:684 +0xa5
```

Error not reproducible locally. Fixed by deleting all keys from redis.

**To Reproduce**

No clear steps to reproduce


**Version**

https://github.com/argoproj/argo-cd/commit/3bba77c36fd2518c02e8757e502b786624f41c2e

Similar but not quite:

```
time="2021-08-10T09:09:43Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=GenerateManifest grpc.request.deadline="2021-08-10T09:10:43Z" grpc.service=repository.RepoServerService grpc.start_time="2021-08-10T09:09:43Z" grpc.time_ms=183.646 span.kind=server system=grpc
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x1f71f66]

goroutine 215 [running]:
github.com/argoproj/argo-cd/v2/reposerver/apiclient.(*ManifestResponse).MarshalToSizedBuffer(0x0, 0x54b8b30, 0x0, 0x0, 0x40bb45, 0x2a208a0, 0x2ce2820)
        /go/src/github.com/argoproj/argo-cd/reposerver/apiclient/repository.pb.go:2043 +0x26
github.com/argoproj/argo-cd/v2/reposerver/apiclient.(*ManifestResponse).Marshal(0x0, 0x2ce2820, 0x0, 0x7f70ef7a53b0, 0x0, 0x1)
        /go/src/github.com/argoproj/argo-cd/reposerver/apiclient/repository.pb.go:2026 +0x7a
google.golang.org/grpc/encoding/proto.codec.Marshal(0x2ce2820, 0x0, 0x0, 0x0, 0x40a51f, 0xc00003a000, 0x2a3a460)
        /go/pkg/mod/google.golang.org/grpc@v1.15.0/encoding/proto/proto.go:70 +0x199
google.golang.org/grpc.encode(0x7f70ef7e7cd8, 0x54b8b30, 0x2ce2820, 0x0, 0x54b8b30, 0x3c4c378, 0xc000575830, 0x2d38040, 0xc0000b0000)
        /go/pkg/mod/google.golang.org/grpc@v1.15.0/rpc_util.go:511 +0x52
google.golang.org/grpc.(*Server).sendResponse(0xc00077e480, 0x3c6fe98, 0xc000fc2f00, 0xc000775f00, 0x2ce2820, 0x0, 0x0, 0x0, 0xc00057c7cb, 0x0, ...)
        /go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:805 +0x91
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00077e480, 0x3c6fe98, 0xc000fc2f00, 0xc000775f00, 0xc0007dae40, 0x535db00, 0x0, 0x0, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:1004 +0x5d1
google.golang.org/grpc.(*Server).handleStream(0xc00077e480, 0x3c6fe98, 0xc000fc2f00, 0xc000775f00, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:1208 +0x122c
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000e820e0, 0xc00077e480, 0x3c6fe98, 0xc000fc2f00, 0xc000775f00)
        /go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:686 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /go/pkg/mod/google.golang.org/grpc@v1.15.0/server.go:684 +0xa5
```

--- a/reposerver/cache/cache.go
+++ b/reposerver/cache/cache.go
@@ -152,9 +152,9 @@ func (c *Cache) GetManifests(revision string, appSrc *appv1.ApplicationSource, c
 		return fmt.Errorf("Unable to generate hash value: %s", err)
 	}
 
-	// If the expected hash of the cache entry does not match the actual hash value...
-	if hash != res.CacheEntryHash {
-		log.Warnf("Manifest hash did not match expected value, treating as a cache miss: %s", appName)
+	// If cached result does not have manifests or the expected hash of the cache entry does not match the actual hash value...
+	if hash != res.CacheEntryHash || res.ManifestResponse == nil && res.MostRecentError == "" {
+		log.Warnf("Manifest hash did not match expected value or cached manifests response is empty, treating as a cache miss: %s", appName)
 
 		err = c.DeleteManifests(revision, appSrc, clusterInfo, namespace, appLabelKey, appName)
 		if err != nil {

--- a/reposerver/repository/repository_test.go
+++ b/reposerver/repository/repository_test.go
@@ -168,6 +168,22 @@ func TestGenerateManifests_K8SAPIResetCache(t *testing.T) {
 	assert.True(t, len(res.Manifests) > 1)
 }
 
+func TestGenerateManifests_EmptyCache(t *testing.T) {
+	service := newService("../..")
+
+	src := argoappv1.ApplicationSource{Path: "manifests/base"}
+	q := apiclient.ManifestRequest{
+		Repo: &argoappv1.Repository{}, ApplicationSource: &src,
+	}
+
+	err := service.cache.SetManifests(mock.Anything, &src, &q, "", "", "", &cache.CachedManifestResponse{ManifestResponse: nil})
+	assert.NoError(t, err)
+
+	res, err := service.GenerateManifest(context.Background(), &q)
+	assert.NoError(t, err)
+	assert.True(t, len(res.Manifests) > 0)
+}
+
 // ensure we can use a semver constraint range (>= 1.0.0) and get back the correct chart (1.0.0)
 func TestHelmManifestFromChartRepo(t *testing.T) {
 	service := newService(".")

